### PR TITLE
added Test_server.py with error in 3

### PR DESCRIPTION
**Pull request by Shivank98**
- [ * ] My title is a short description of what I am trying to do
- [ ] I have mentioned all the issues I am fixing
- [ ] Proposed changes

I am facing trouble in creating a third test function.

### CODE REVIEW

**Comment by Shivank98** [comment]
* Facing following error.

```
 test_file_upload _______________________________

    def test_file_upload():
        files= '/question/1/output.txt'
        global url
        url += '/check/1/output.txt'
>       assert requests.post(url, files = files).status_code == 200   # givin a error

test_server.py:19:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ __ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ __ _
../../../anaconda3/lib/python3.7/site-packages/requests/api.py:116: in post
    return request('post', url, data=data, json=json, **kwargs)
../../../anaconda3/lib/python3.7/site-packages/requests/api.py:60: in request
    return session.request(method=method, url=url, **kwargs)
../../../anaconda3/lib/python3.7/site-packages/requests/sessions.py:510: inrequest
    prep = self.prepare_request(req)
../../../anaconda3/lib/python3.7/site-packages/requests/sessions.py:453: inprepare_request
    hooks=merge_hooks(request.hooks, self.hooks),
../../../anaconda3/lib/python3.7/site-packages/requests/models.py:316: in prepare
    self.prepare_body(data, files, json)
../../../anaconda3/lib/python3.7/site-packages/requests/models.py:504: in prepare_body
    (body, content_type) = self._encode_files(files, data)
../../../anaconda3/lib/python3.7/site-packages/requests/models.py:126: in _encode_files
    files = to_key_val_list(files or {})
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ __ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ __ _

value = '/question/1/output.txt'

    def to_key_val_list(value):
        """Take an object and test tosee if it can be represented as a
        dictionary. If it can be, return a list of tuples, e.g.,

        ::

            >>> to_key_val_list([('key', 'val')])
            [('key', 'val')]
            >>> to_key_val_list({'key': 'val'})
            [('key', 'val')]
            >>> to_key_val_list('string')
            ValueError: cannot encodeobjects that are not 2-tuples.

        :rtype: list
        """
        if value is None:
            return None

        if isinstance(value, (str, bytes, bool, int)):
>           raise ValueError('cannot encode objects that are not 2-tuples')
E           ValueError: cannot encodeobjects that are not 2-tuples

../../../anaconda3/lib/python3.7/site-packages/requests/utils.py:301: ValueError
====================== 1 failed, 2 passed in 0.90 seconds ======================
```

**Comment by theSage21** [comment]
I think you have to open an actual file. not pass a string

**Comment by Shivank98** [comment]
> I think you have to open an actual file. not pass a string
oh, ok

**Comment by Shivank98** [comment]
now a new error.  i have no idea what's going on. i hope it's not irritating.
```
platform linux -- Python 3.7.2, pytest-3.8.0, py-1.6.0, pluggy-0.7.1
rootdir: /home/shivank98/Desktop/PyJudge/PyJudge, inifile:
plugins: remotedata-0.3.0, openfiles-0.3.0, doctestplus-0.1.3, arraydiff-0.2, celery-4.2.1
collected 3 items

test_server.py ..F                                  [100%]

======================== FAILURES =========================
____________________ test_file_upload _____________________

    def test_file_upload():
        files= open('files/questions/1/output.txt')
        global url
        url += '/check/1/output.txt'

>       assert requests.post(url, files = files).status_code == 200   # givin a error

test_server.py:20:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../../anaconda3/lib/python3.7/site-packages/requests/api.py:116: in post
    return request('post', url, data=data, json=json, **kwargs)
../../../anaconda3/lib/python3.7/site-packages/requests/api.py:60: in request
    return session.request(method=method, url=url, **kwargs)
../../../anaconda3/lib/python3.7/site-packages/requests/sessions.py:510: in request
    prep = self.prepare_request(req)
../../../anaconda3/lib/python3.7/site-packages/requests/sessions.py:453: in prepare_request
    hooks=merge_hooks(request.hooks, self.hooks),
../../../anaconda3/lib/python3.7/site-packages/requests/models.py:316: in prepare
    self.prepare_body(data, files, json)
../../../anaconda3/lib/python3.7/site-packages/requests/models.py:504: in prepare_body
    (body, content_type) = self._encode_files(files, data)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

files = ['-1103\n', '32848\n', '-70788\n', '49859\n', '70551\n', '-43971\n', ...]
data = {}

    @staticmethod
    def _encode_files(files, data):
        """Build the body for a multipart/form-data request.

            Will successfully encode files when passed as a dict or a list of
            tuples. Order is retained if data is a list oftuples but arbitrary
            if parameters are supplied as a dict.
            The tuples may be 2-tuples (filename, fileobj), 3-tuples (filename, fileobj, contentype)
            or 4-tuples (filename, fileobj, contentype, custom_headers).
            """
        if (not files):
            raise ValueError("Files must be provided.")
        elif isinstance(data, basestring):
            raise ValueError("Data must not be a string.")

        new_fields = []
        fields = to_key_val_list(data or {})
        files = to_key_val_list(files or {})

        for field, val in fields:
            if isinstance(val, basestring) or not hasattr(val, '__iter__'):
                val = [val]
            for v in val:
                if v is not None:
                    # Don't call str() on bytestrings: in Py3 it all goes wrong.
                    if not isinstance(v, bytes):
                        v = str(v)

                    new_fields.append(
                        (field.decode('utf-8') if isinstance(field, bytes) else field,
                         v.encode('utf-8') if isinstance(v, str) else v))

>       for (k, v) in files:
E       ValueError: too many values to unpack (expected 2)

../../../anaconda3/lib/python3.7/site-packages/requests/models.py:141: ValueError
=========== 1 failed, 2 passed in 0.74 seconds ============
```

File: test_server.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+def test_question():
+    r = requests.get(url+'/question/1')
+    assert r.status_code == 200
+
+def test_download():
+    assert requests.get(url+'/question/1/inputs.txt').status_code == 200
+
+def test_file_upload():
+    files= open('files/questions/1/output.txt')
```
Comment (theSage21) [inline test_server.py]: Why are you doing this?

Comment (theSage21) [reply test_server.py]: Actually tell me how you decided to write this line. Maybe I can help you then.

Comment (Shivank98) [reply test_server.py]: i have no idea what i thought then, i just wanted to create a test function which can verify that file is posted.

**Comment by Shivank98** [comment]
> Yaar [jo document](http://docs.python-requests.org/en/master/user/quickstart/#post-a-multipart-encoded-file) padhne ko bola tha woh padha hai?

yes, i have read that.

File: test_server.py (hunk 1 above)
Comment (theSage21) [reply test_server.py]: let me rephrase.

What steps did you follow to write that test function. Here's what I usually do:

1. google for how to upload file
2. find documentation and read it
3. try to find an exact example of what I'm trying to do
4. understand the parts of the example that are relevant
5. copy paste into my code

Now. What did you do?

Comment (theSage21) [reply test_server.py]: for example, if you followed a similar list, how did you find the example of what you are trying to do?

Comment (Shivank98) [reply test_server.py]: I did the exact above things, I was not able to find out the exact example.
but as you say I will do it once again and will discuss the example with you.
sounds cool?

Comment (theSage21) [reply test_server.py]: I think I linked to the exact place in the document which we needed to implemented?
image

Comment (Shivank98) [reply test_server.py]:
```
def test_file_upload():
    files= {'file' : open('files/questions/1/output.txt', 'rb')}
    global url
    url += '/check/1/output.txt'

    assert requests.post(url, files = files).status_code == 200   # givin a error
```

is this correct?

Comment (theSage21) [reply test_server.py]: is it running?

Comment (Shivank98) [reply test_server.py]: no, this is what i am getting.
```
platform linux -- Python 3.7.2, pytest-3.8.0, py-1.6.0, pluggy-0.7.1
rootdir: /home/shivank98/Desktop/PyJudge/PyJudge, inifile:
plugins: remotedata-0.3.0, openfiles-0.3.0, doctestplus-0.1.3, arraydiff-0.2, celery-4.2.1
collected 3 items

test_server.py ..F                                  [100%]

======================== FAILURES =========================
____________________ test_file_upload _____________________

    def test_file_upload():
        files= {'file' : open('files/questions/1/output.txt', 'rb')}
        global url
        url += '/check/1/output.txt'

>       assert requests.post(url, files = files).status_code == 200   # givin a error
E       AssertionError: assert 404 == 200
E        +  where 404 = <Response [404]>.status_code
E        +    where <Response [404]> = <function post at 0x7fd95ef0b268>('http://localhost:8080/check/1/output.txt', files={'file': <_io.BufferedReader name='files/questions/1/output.txt'>})
E        +      where <function post at 0x7fd95ef0b268> = requests.post

test_server.py:20: AssertionError
```

Comment (theSage21) [reply test_server.py]: please debug. The error message is on your screen. I can see the mistake but it might be better if you find it on your own.

Comment (Shivank98) [reply test_server.py]: please just tell me, so I can just end this issue and start working from fresh on a new issue. i know i can do it my own too. but ...

Comment (theSage21) [reply test_server.py]: No.

The point is for you to learn. You have to go to the gym and lift your own weights. If someone else does it for you, going to the gym is useless.

Take your time. There is no time limit. Do it right.

Comment (Shivank98) [reply test_server.py]: Ohk,

Comment (theSage21) [reply test_server.py]: [Dashrath Manjhi](https://en.wikipedia.org/wiki/Dashrath_Manjhi) ko yaad rakho. Pahaad toda tha unhone! 20 saal lage the. Is issue ko toh sirf ek hafta hua hai :smile: Lage raho janaab, jo gyaan ka phool uge ga na aap mein aage ki zindagi aasaan ho jaayegi

Comment (Shivank98) [reply test_server.py]: give me some hints so I can do it asap.

Comment (Shivank98) [reply test_server.py]: @theSage21 i can see the given url is incorrect.
for `url += '/question/1'` i am receiving the `405`

Comment (rishabhKalakoti) [reply test_server.py]: @Shivank98 Shouldn't the check url be `/check/1` instead of `/check/1/output.txt`
According to me the program is giving 404 for not finding the handler for `/check/1/output.txt`

Comment (Shivank98) [reply test_server.py]: code
```
def test_file_upload():
    file= {'file' : open('files/questions/1/output.txt', 'rb')}
    
    global url
    url += '/check/1'
    response = requests.post(url, data=file)
    assert response.status_code == 200
```
error
```
def test_file_upload():
        file= {'file' : open('files/questions/1/output.txt', 'rb')}

        global url
        url += '/check/1'
        response = requests.post(url, data=file)
>       assert response.status_code == 200
E       assert 413 == 200
E        +  where 413 = <Response [413]>.status_code

test_server.py:26: AssertionError
```

Comment (theSage21) [reply test_server.py]: would you like to pick up some other issue? Maybe someone else can do this?

Comment (rishabhKalakoti) [reply test_server.py]: Got it!
Use,
```
def test_file_upload():
    files= {'upload' : open('files/questions/1/output.txt', 'rb')}
    global url
    url += '/check/1'
    assert requests.post(url, files = files).status_code == 200

```

Comment (Shivank98) [reply test_server.py]: damn it, is it what I was missing.?
fool me.

**Comment by Shivank98** [comment]
@theSage21 @rishabhKalakoti  finally it happened. I could not see what I was missing, though it was an easy thing. just because I didn't try to go back and see the function code it took a lot of time.
Though leaned a lesson.

**Comment by Shivank98** [comment]
@theSage21 is it fine.
time to merge?

File: test_server.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    #time = datetime.datetime.now()
+
+    # type(uploaded) == <class 'bytes'>
+    # uploaded outputs by user
+    assert requests.get(url+'/question/1/inputs.txt').status_code == 200
+
+def test_file_upload():
+    files= {'upload' : open('files/questions/1/output.txt', 'rb')}
+    global url
+    url += '/check/1'
```
Comment (theSage21) [inline test_server.py]: this alters `url`
```suggestion
    check_url = url + '/check/1'
```

**Comment by Shivank98** [comment]
working fine.
